test/box/mvcc_mass.test.lua

```diff
@@ -0,0 +1,423 @@
+    for k, v in pairs(rhs) do
+        if not is_equal(lhs[k], v) then return false end
+    end
+
+    return true
+end;
+
+function try_apply_txn(stmts)
+    local nstmts = #stmts
+    for i = 1, nstmts do
+        local stmt = stmts[i]
+        local res = yaml.decode(console.eval('box.space.s:'..stmt.op))
+        if not is_equal(res, stmt.res) then
+            return false, 'TXN #'..stmt.t..' OP '..stmt.op..
+                          ' FAILED: expected '..json.encode(res)..
+                          ' actual '..json.encode(stmt.res)
+        end
+    end
+
+    return true, 'OK'
+end;
+
+function try_serialize(stmts, ntx, ro_mask, ok_mask)
+    local rw_seq = {}
+    local ro_set = {}
+    local txn_stmts = {}
+    for _ = 1, ntx do
+        table.insert(txn_stmts, {})
+    end
+
+    local nstmts = #stmts
+    for i = 1, nstmts do
+        local stmt = stmts[i]
+        if stmt.op == 'begin' and ok_mask[stmt.t] and ro_mask[stmt.t] then
+            table.insert(ro_set, stmt.t)
+        end
+        if stmt.op == 'commit' and ok_mask[stmt.t] and not ro_mask[stmt.t] then
+            table.insert(rw_seq, stmt.t)
+        end
+
+        if stmt.op ~= 'begin' and stmt.op ~= 'commit' and
+            stmt.op ~= 'rollback' then
+            table.insert(txn_stmts[stmt.t], stmt)
+        end
+    end
+
+    local nro_ok = 0
+    local nro = #ro_set
+-- Trying to serialize ROs first
+    for i, s in pairs(ro_set) do
+        if try_apply_txn(txn_stmts[s]) then
+            nro_ok = nro_ok + 1
+            ro_set[i] = nil
+        end
+    end
+
+    local nrw_seq = #rw_seq
+    for k = 1, nrw_seq do
+        local t = rw_seq[k]
+-- Trying to serialize RW
+        local result = true
+        local what = 'OK'
+        result, what = try_apply_txn(txn_stmts[t])
+        if result == false then
+            return result, what
+        end
+-- Trying to serialize ROs after
+        for i, s in pairs(ro_set) do
+            if try_apply_txn(txn_stmts[s]) then
+                nro_ok = nro_ok + 1
+                ro_set[i] = nil
+            end
+        end
+    end
+
+    if nro ~= nro_ok then
+        local ro_bad = {}
+        for _, t in pairs(ro_set) do
+            table.insert(ro_bad, t)
+        end
+        return false, 'RO TXNS FAILED TO SERIALIZE '..json.encode(ro_bad)
+    end
+
+    return true, 'OK'
+end;
+
+test_run:cmd('setopt delimiter ""');
+-- END GUEST API DEFINITIONS
+--------------------------------------------------------------------------------
+
+-- START CONNECTION TO HOST
+conn = netbox.connect('127.0.0.1:3333')
+-- conn = netbox.connect(os.getenv("LISTEN"))
+conn:wait_connected()
+
+seq = box.schema.space.create('s', {engine='memtx', if_not_exists=true})
+seq_tree = seq:create_index('st', {type='tree'})
+seq_hash = seq:create_index('sh', {type='hash'})
+
+ntx, nro = 20, 8
+niters = 500
+stmts, ro_mask, ok_mask = conn:call('gen_statements', {ntx, nro, niters})
+result, what = try_serialize(stmts, ntx, ro_mask, ok_mask)
```

Comment:
I think we should do `gen_statements` + `try_serialize` in cycle, recreating spaces each time.

Reply:
Fixed.